src/ListClient.ts

```diff
@@ -48,10 +59,21 @@ export class InnerListClient<T extends any> implements ObjectClient {
   }
 
   private sendCmd(cmd: string[]) {
+    if (this.throwsOnMutate) {
+      throw errNoInfiniteLoop();
+    }
+
     this.ws.send('doc:cmd', {
       room: this.roomID,
       args: cmd,
     });
+
+    this.throwsOnMutate = true;
+    this.bus.publish({
+      args: cmd,
+      from: this.actor,
+    });
+    this.throwsOnMutate = false;
   }
 
   private clone(): InnerListClient<T> {
```

Comment:
I wonder if `clone()` could create issues since it is copying the `throwsOnMutate` member variable.

`clone()` is called in `dangerouslyUpdateInnerClient` and the new object is returned, eventually being passed to the subscribed callback. The original copy of the client clears its `throwsOnMutate` flag, but the newly-cloned client passed into the subscribe callback will never have the flag cleared AFAICT. I think this makes the example fix in the error markdown invalid, as the `newState` client will have `throwsOnMutate` already set.

I can't think of a great way to fix this off-hand. The flag being set on the client passed into the callback is what is protecting the user from mutating in the callback. What if we just didn't `clone()`? I'm actually not sure why it's necessary currently. If we need some fields cloned, maybe `throwsOnMutate` can be a ref that is shared between all of the cloned clients?

Reply:
YUP. Added a test to catch this.

> What if we just didn't clone()? I'm actually not sure why it's necessary currently.

The reason we clone here is because React / Vue / Svelte won't rerender if it thinks the object is the same. Cloning forces a new reference and tricks other libraries into re-rendering when stuff's changed.

So if we didn't clone, this wouldn't work:
```jsx
function useMap() {
  const [map, setMap] = useState()

  // ... later
  setMap(map.set("dogs", "cats")) // React needs the result here to be a new reference 
} 
```

> I can't think of a great way to fix this off-hand. [...] maybe throwsOnMutate can be a ref that is shared between all of the cloned clients?

I'm going to try something like the ref shared between the clients; I'll just put the loop checking into the `LocalBus`, which itself is an object that won't be recreated when we clone.